anndata/tests/test_helpers.py

```diff
@@ -246,3 +248,32 @@ def test_assert_equal_dask_sparse_arrays():
 
     assert_equal(x, y)
     assert_equal(y, x)
+
+
+@pytest.mark.parametrize(
+    "error, match",
+    [
+        (Exception("test"), "test"),
+        (add_note(AssertionError("foo"), "bar"), "bar"),
+        (add_note(add_note(AssertionError("foo"), "bar"), "baz"), "bar"),
+        (add_note(add_note(AssertionError("foo"), "bar"), "baz"), "baz"),
+    ],
+)
+def test_check_error_notes_success(error, match):
+    with pytest.raises(Exception) as exc_info:
+        raise error
+    check_error_or_notes_match(exc_info, match)
```

Comment:
This pattern is difficult to understand. I think our helper should be a context manager instead, replacing `pytest.raises`. The odd use of a context manager `as` value after the end of the block is fine, but as a primary pattern, it’s confusing.

Reply:
* This is how the [pytest docs say to use their context manager](https://docs.pytest.org/en/7.4.x/reference/reference.html#pytest.raises)
* I'd like to see if we can get pytest support for this: https://github.com/pytest-dev/pytest/issues/11223
* Some of these tests already do multiple checks, so this wouldn't change the usage of the pattern

Reply:
bullet point 1&3 are covered by

> The odd use of a context manager as value after the end of the block is fine, but as a primary pattern, it’s confusing.

Yes, bullet point 2 would be ideal. Until then, we should create a nice API.

Reply:
That doesn’t contradict what I want. Maybe I should be more explicit. I’m thinking of a trivial solution like this: 
no complexity, mirrors the pytest API.

```py
from contextlib import contextmanager

@contextmanager
def raises_note(cls, *, matches_message: str | None = None, matches_notes: str | None = None) -> pytest.ExceptionInfo:
    with pytest.raises(cls, matches=matches_message) as exc_info:
        yield exc_info
    assert any(re.search(matches_notes, note) for note in e.value.__notes__)
```

(didn’t test, but it shouldn’t be much more than that)